File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        
        if not root:
            return False
        
        queue = deque()
        queue.append([root, root.val])
        visited = set()
        visited.add(root)
        answer = False

        while queue:
            n,  sum_n = queue.popleft()
            if not n.left and not n.right:
                if sum_n == targetSum:
                    answer |= True
                answer |= False

            for child in [n.left, n.right]:
                if child and child not in visited:
                    queue.append([child, sum_n + child.val])
                    visited.add(child)
        
        return answer

File: 0112-path-sum/0112-path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, targetSum int) bool {
    
    if root == nil{
        return false
    }

    if root.Left == nil && root.Right == nil{
        return root.Val == targetSum
    }

    left_sum := hasPathSum(root.Left, targetSum - root.Val)
    right_sum := hasPathSum(root.Right, targetSum - root.Val)

    return left_sum || right_sum
}